atlasdb-config/src/main/java/com/palantir/atlasdb/http/BlockingTimeoutExceptionMapper.java

```diff
@@ -15,14 +15,37 @@
  */
 package com.palantir.atlasdb.http;
 
+import java.time.Duration;
+
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
 import javax.ws.rs.core.Response;
 import javax.ws.rs.ext.ExceptionMapper;
 
+import com.palantir.conjure.java.api.errors.QosException;
 import com.palantir.lock.remoting.BlockingTimeoutException;
 
+/**
+ * Converts {@link BlockingTimeoutException}s into appropriate status responses, depending on the user's
+ * {@link AtlasDbHttpProtocolVersion}. The intention is that clients should retry on the same node (as in the absence
+ * of exceptional circumstances, it would still be the leader), and they may do so immediately (as an individual lock
+ * being locked does not imply that the server is struggling).
+ *
+ * This is a 503 without a Retry-After header and with a message body corresponding to {@link BlockingTimeoutException}
+ * in {@link AtlasDbHttpProtocolVersion#LEGACY_OR_UNKNOWN}.
+ */
 public class BlockingTimeoutExceptionMapper implements ExceptionMapper<BlockingTimeoutException> {
```

Comment:
As much as I hate to say it, let's have a abstract class:

```java
abstract class ProtocolAwareExceptionMapper<T> implements ExceptionMapper<T> {

  private final HttpProtocolAwareExceptionTranslator<T> translator;
  @Context private HttpHeaders httpHeaders;
  
  public ProtocolAwareExceptionMapper(AtlasDbHttpProtocolHandler<T> handler) {
    ...
  }

  @Override
  public Response toResponse(T exception) {
    return translator.translate(httpHeaders, exception);
  }
} 
```

with impls:

```java
public class BlockingTimeoutExceptionMapper extends ProtocolAwareExceptionMapper<BlockingTimeoutException> {
  BlockingTimeoutExceptonMapper() {
     super(LambdaHandler.of(...));
  }
}
```

In an ideal world, we'd just create the exception mapper and pass in the protocol handler ourselves, but I'm not sure how that plays with the `@Context` injection. If it *does* work, I'd prefer it this way - that is without the abstract class.

Reply:
I think that means we can get rid of the `HttpProtocolAwareExceptionTranslator` and just inline its implementation, what do you think?

Reply:
Yep, will test this out. As discussed this design would be good (if it works)